src/data_constructors.jl

```diff
@@ -89,35 +38,190 @@ and present inside `data`.
-    ts = apply_schema(formula.rhs, sch)
-    ts = collect_matrix_terms(ts)
-    X = modelcols(ts, Tables.columntable(data))
+    if has_ranef(formula)
+        X = MixedModels.modelmatrix(MixedModel(formula, data))
+        X = X[:, 2:end]
+    else
+        X = StatsModels.modelmatrix(formula, data)
+        if hasintercept(formula)
+            X = X[:, 2:end]
+        end
+    end
     return X
 end
 
 """
     data_random_effects(formula::FormulaTerm, data)
 
-Constructs the matrix(ces) Z(s) of random-effects (a.k.a. group-level) predictors.
+Constructs the vector(s)/matrix(ces) Z(s) of random-effects (a.k.a. group-level)
+slope predictors.
 
-Returns a Tuple of `Matrix` of the random-effects predictors variables in the `formula`
-and present inside `data`.
+Returns a `Dict{String, AbstractArray}` of `Vector`/`Matrix` as values of the random-effects
+predictors slope variables (keys) in the `formula` and present inside `data`.
 
 # Arguments
 - `formula`: a `FormulaTerm` created by `@formula` macro.
 - `data`:  a `data` object that satisfies the
 [Tables.jl](https://github.com/JuliaData/Tables.jl) interface such as a DataFrame.
 """
 function data_random_effects(formula::FormulaTerm, data::D) where {D}
-    Tables.istable(data) || throw(ArgumentError("Data of type $D is not a table!"))
-    Z = nothing
-    # TODO:
-    # is_matrix_terms false for random-effects.
+    # with zerocorr we create only vectors and add them one by one with NCP
+    # without zerocorr we create a full-blown matrix with NCP
+    Z = Dict{String,AbstractArray}() # empty Dict
+    if !has_ranef(formula)
+        return nothing
+    end
+    slopes = slope_per_ranef(ranef(formula))
+
+    if length(slopes) > 0
+        # add the slopes to Z
+        # this would need to create a vector from the column of the X matrix from the
+        # slope term
+        for slope in values(slopes.grouping_vars)
+            if slope isa String
+                Z["slope_" * slope] = get_var(term(slope), data)
+            else
+                for s in slope
+                    Z["slope_" * s] = get_var(term(s), data)
+                end
+            end
+        end
+    else
+        Z = nothing
+    end
     return Z
 end
 
-function _has_ranef(formula::FormulaTerm, data)
-    return false
+"""
+    has_ranef(formula::FormulaTerm)
+
+Returns `true` if any of the terms in `formula` is a `FunctionTerm` or false
+otherwise.
+"""
+function has_ranef(formula::FormulaTerm)
+    return any(t -> t isa FunctionTerm, formula.rhs)
+end
+
+"""
+    ranef(formula::FormulaTerm)
+
+Returns a tuple of the `FunctionTerm`s parsed as `RandomEffectsTerm`s in `formula`.
+If there are no `FunctionTerm`s in `formula` returns `nothing`.
+"""
+function ranef(formula::FormulaTerm)
+    if has_ranef(formula)
+        terms = filter(t -> t isa FunctionTerm{typeof(|)}, formula.rhs)
+        terms = map(
+            t -> RandomEffectsTerm(first(t.args_parsed), last(t.args_parsed)), terms
+        )
```

Comment:
Maybe change to something like

```suggestion
terms = map(terms) do t
     lhs, rhs = first(t.args_parsed), last(t.args_parsed)
     RandomEffectsTerm(u, v)
end
```

for readability

Reply:
WTF with the `u, v`? I've corrected it 😂